pyvelox/complex.cpp

```diff
@@ -0,0 +1,160 @@
+namespace py = pybind11;
+
+struct ElementCounter {
+  vector_size_t insertedElements =
+      0; // to track the elements already in the vector
+  vector_size_t totalElements = 0;
+  std::vector<ElementCounter> children;
+};
+
+void checkOrAssignType(TypePtr& type, const std::function<TypePtr()>& func) {
+  if (type->kind() == TypeKind::UNKNOWN) {
+    type = func();
+  } else if (!(type->kindEquals(func()))) {
+    throw py::type_error(
+        "Cannot construct type tree, invalid variant for complex type");
+  }
+}
+
+template <TypeKind Kind>
+void setElementInFlatVector(
+    vector_size_t idx,
+    const variant& v,
+    VectorPtr& vector) {
+  using NativeType = typename TypeTraits<Kind>::NativeType;
+  auto asFlat = vector->asFlatVector<NativeType>();
+  asFlat->set(idx, NativeType{v.value<NativeType>()});
+}
+
+void constructType(TypePtr& type, const variant& v, ElementCounter& counter) {
+  ++counter.totalElements;
+
+  if (v.isNull()) {
+    // since the variant is NULL, we can't infer the data type
+    // of it's infer, thus it maybe UNKNOWN or INVALID at this stage
+    // which implies further investigation is required
+    if (v.kind() != TypeKind::UNKNOWN && v.kind() != TypeKind::INVALID &&
+        v.kind() != type->kind()) {
+      throw std::invalid_argument("Variant was of an unexpected kind");
+    }
+    return;
+  } else {
+    // if a Non-Null variant's type is unknown or not one of the valid
+    // types which are supported then the Type tree cannot be constructed
+    if (v.kind() == TypeKind::UNKNOWN || v.kind() == TypeKind::INVALID) {
+      throw std::invalid_argument(
+          "Non-null variant has unknown or invalid kind");
+    }
+
+    switch (v.kind()) {
+      case TypeKind::ARRAY: {
+        counter.children.resize(1);
+        auto asArray = v.array();
+        TypePtr childType = createType(TypeKind::UNKNOWN, {});
+        for (const auto& element : asArray) {
+          constructType(childType, element, counter.children[0]);
+        }
+
+        // if child's type still remains Unknown, implies all the
+        // elements in the array are actually NULL
+        if (childType->kind() == TypeKind::UNKNOWN) {
+          throw py::value_error("Cannot construct array with all None values");
+        }
+        checkOrAssignType(type, [&childType]() {
+          return createType<TypeKind::ARRAY>({childType});
+        });
+        break;
+      }
+
+      default: {
+        checkOrAssignType(type, [&v]() { return createScalarType(v.kind()); });
+        break;
+      }
+    }
+  }
+}
+
+static void insertVariantIntoVector(
+    TypeKind& typeKind,
+    const variant& v,
+    VectorPtr& vector,
+    ElementCounter& counter) {
+  if (v.isNull()) {
+    vector->setNull(counter.insertedElements, true);
+  } else {
+    switch (typeKind) {
+      case TypeKind::ARRAY: {
+        auto asArray = vector->as<ArrayVector>();
+        asArray->elements()->resize(counter.children[0].totalElements);
```

Comment:
nit: as an aside I am not really a fan of using one struct which has information on number of total elements also being mutated for elements inserted. This is because there is no way we can maintain constness on totalElements etc. Maybe we should split up that structure ?

Reply:
Apart from the totalElements and the insertedElements, it is the children vector that tracks the number of types of elements the complex vector holds. Splitting the struct will result in two different struct or entities being tracked for the children and the elements inserted/total contained.